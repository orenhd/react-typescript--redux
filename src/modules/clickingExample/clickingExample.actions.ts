import { Action, ActionCreator } from 'redux';

import { State } from '../../application/application.store';

import * as dataModels from './clickingExample.dataModels';

export const SET_USER_NAME: string = 'SET_USER_NAME';
export const setUserName:ActionCreator<Action> = (userName: string) => {
    return {
        type: SET_USER_NAME,
        userName
    }
}

export const HOME_BUTTON_CLICKED: string = 'HOME_BUTTON_CLICKED';
export const homeButtonClicked:ActionCreator<Action> = () => {
    return {
        type: HOME_BUTTON_CLICKED,
    }
}

export const HOME_BUTTON_CLICKED_OUTSIDE: string = 'HOME_BUTTON_CLICKED_OUTSIDE';
export const homeButtonClickedOutside:ActionCreator<Action> = () => {
    return {
        type: HOME_BUTTON_CLICKED_OUTSIDE
    }
}