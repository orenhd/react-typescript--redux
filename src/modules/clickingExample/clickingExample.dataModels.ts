export interface ClickingData {
    homeButtonClickCount: number;
    homeButtonClickOutsideCount: number;
}