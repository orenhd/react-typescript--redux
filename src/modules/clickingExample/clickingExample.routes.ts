import { RouteConfig } from 'vue-router';

import ClickingExample from "./clickingExample.vue";

const routes: RouteConfig[] =  [
    { path: '/clicking-example', component: ClickingExample }
];

export default routes;