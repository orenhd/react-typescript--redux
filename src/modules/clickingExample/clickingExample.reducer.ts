import { Action } from 'redux';

import { SET_USER_NAME, UPDATE_CLICKING_DATA } from './clickingExample.actions';

import * as dataModels from './clickingExample.dataModels';

export interface state {
  userName: string;
  clickingData: dataModels.ClickingData;
}

const initialState: state = {
  userName: 'World',
  clickingData: {}
}

export function reducer(state: state = initialState, action: any) {
  switch (action.type) {
    case SET_USER_NAME:
      return { ...state, userName: action.userName };
    case UPDATE_CLICKING_DATA:
      return { ...state, clickingData: action.clickingData };
    default:
      return state;
  }
}