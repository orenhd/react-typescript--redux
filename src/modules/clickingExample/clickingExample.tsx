import React, { PureComponent } from "react";
import { Dispatch, bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { State } from '../../application/application.store' 
import { setUserName, updateClickingData } from './clickingExample.actions';

import * as dataModels from './clickingExample.dataModels';

import UserNameBar from './components/userNameBar';
import ClickingPanel from './components/clickingPanel';

interface ClickingExmapleProps {
    userName: string;
    clickingData: dataModels.ClickingData;
    setUserName: any;
    updateClickingData: any;
}

class ClickingExample extends PureComponent<ClickingExmapleProps, {}> {

    /* Class Methods */

    setUserName = (userName: string) => {
        this.props.setUserName(userName);
    }

    homeButtonClicked = () => {
        this.props.updateClickingData(dataModels.ClickCountTypes.homeButtonClick);
    }

    homeButtonClickedOutside = () => {
        this.props.updateClickingData(dataModels.ClickCountTypes.homeButtonClickOutside);
    }

    render() {
        return <div className="clicking-example margined-content">
            <UserNameBar 
                userName={this.props.userName} 
                userNameChangedHandler={this.setUserName}
            />
            <ClickingPanel
                clickingData={this.props.clickingData}
                homeButtonClickedHandler={this.homeButtonClicked}
                homeButtonClickedOutsideHandler={this.homeButtonClickedOutside}
            />
        </div>
    }
}

const mapStateToProps = (state: State) => {
    const { clickingExample } = state;
    const { userName, clickingData } = clickingExample;
    return {
        userName,
        clickingData
    }
}

const mapDispatchToProps = (dispatch: Dispatch<State>) => {
    return bindActionCreators({
        setUserName,
        updateClickingData
    }, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(ClickingExample);