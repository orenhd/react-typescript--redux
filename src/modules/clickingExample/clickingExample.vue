<template>
    <div class="clicking-example mx-3 my-3">
        <user-name-bar
            :userName="userName$"
            @userNameChanged="setUserName($event)"
        >
        </user-name-bar>
        <clicking-panel
            :clickingData="clickingData$"
            @homeButtonClicked="homeButtonClicked()"
            @homeButtonClickedOutside="homeButtonClickedOutside()"
        >
        </clicking-panel>
    </div>
</template>

<script lang="ts">
import { Vue, Component, Prop } from "vue-property-decorator";
import { Observable } from 'rxjs/Observable';

import UserNameBar from "./components/userNameBar.vue";
import ClickingPanel from "./components/clickingPanel.vue";

import ClickOutside from "../../shared/directives/clickOutside";

import * as clickingExampleService from "./clickingExample.service";

import * as dataModels from "./clickingExample.dataModels";

@Component({components: {UserNameBar, ClickingPanel}, subscriptions: () => {
    return {
        userName$: clickingExampleService.userName$,
        clickingData$: clickingExampleService.clickingData$
    }
}})
export default class ClickingExample extends Vue {

    /* Component Methods */

    setUserName(userName: string) {
        clickingExampleService.setUserName(userName);
    }

    homeButtonClicked() {
        clickingExampleService.homeButtonClicked();
    }

    homeButtonClickedOutside() {
        clickingExampleService.homeButtonClickedOutside();
    }
}
</script>

<style lang="scss" scoped>
    @import '~styles/variables'; //tilde to resolve 'like a module'
</style>
