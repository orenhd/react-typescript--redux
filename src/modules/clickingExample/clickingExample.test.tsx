import React from 'react';
import { shallow } from 'enzyme';

import { ClickingExample } from './clickingExample';

import * as dataModels from './clickingExample.dataModels';

describe('clickingExample', () => {
  it('should render correctly', () => {

    const clickingData: dataModels.ClickingData = {
      [dataModels.ClickCountTypes.homeButtonClick]: 3,
      [dataModels.ClickCountTypes.homeButtonClickOutside]: 4,
    }

    const component = shallow(<ClickingExample
      userName='Worlds'
      clickingData={clickingData}
      setUserName={jest.fn()}
      updateClickingData={jest.fn()}
    />);

    expect(component).toMatchSnapshot();
  });
});