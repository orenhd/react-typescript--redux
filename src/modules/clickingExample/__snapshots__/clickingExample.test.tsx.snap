// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clickingExample should render correctly 1`] = `
<div
  className="clicking-example margined-content"
>
  <UserNameBar
    userName="Worlds"
    userNameChangedHandler={[MockFunction]}
  />
  <ClickingPanel
    clickingData={
      Object {
        "0": 3,
        "1": 4,
      }
    }
    homeButtonClickedHandler={[Function]}
    homeButtonClickedOutsideHandler={[Function]}
  />
</div>
`;
